### DISCUSSION TITLE
Why does the TL select SELF and KNEM when communicating between nodes

### DISCUSSION THREAD

**Discussion by w7658610** [openucx/ucx#8247] [Q&A] [answered]
Hi,

like this, When I used OSU_BW to test communication between 2 nodes (1 process per node).
Why  self and knem(or cma)? Is there an IPC?

![2d9911ba16cc8bedb0007430446abc6](https://user-images.githubusercontent.com/37858414/169448182-aabfac79-310a-48f4-82c3-ff49c077beb0.png)

>[1653018173.525616] [j13r2n01:17472:0]     ucp_worker.c:1627 UCX  INFO  ep_cfg[0]: tag(self/memory knem/memory rc_mlx5/mlx5_0:1 rc_mlx5/mlx5_1:1 rc_mlx5/mlx5_2:1); 
>[1653018173.577027] [j13r2n01:17472:0]     ucp_worker.c:1627 UCX  INFO  ep_cfg[1]: tag(rc_mlx5/mlx5_0:1 rc_mlx5/mlx5_1:1 rc_mlx5/mlx5_2:1 rc_mlx5/mlx5_3:1);

Thanks.

**Comment by brminich** [comment, marked as answer]
I suppose this configuration `ep_cfg[0]` represents a connection to self (the same rank), while `ep_cfg[1]` corresponds to the other rank connection

**Reply by w7658610** [reply]
I still have some questions.

1. What does rank mean here? The process? 

2. And  communication between 2 nodes (1 process per node),why  ep needs " a connection to self  " .

3. Then I notice that self/memory  is usually in `ep_cfg[0]` and posix/memory is in `[1]`.
  - `ep_cfg[0, 1, 2, 3]`，0 1 2 3 ,Is there a relationship between configuration and numbers?

![0adc4aff1c78a15ecc84f20e169ec3f](https://user-images.githubusercontent.com/37858414/169474216-a50f7db2-15dd-4763-a16c-adc8e185ecee.png)

**Reply by hoopoepg** [reply]
1. rank is MPI unit, it may be represented as separate process or even thread in process
2. MPI allows to send messages to itself, for this case UCX creates configuration to send messages like loopback
3. there is no strong correlation in configuration ordering

**Reply by brminich** [reply]
1. You may consider rank as a process (rank is MPI term).
2. In MPI there is a mode when all processes establish connections to all others during MPI_Init (typically this is the case when number of ranks is relatively low)
3. The configuration numbers are just internal UCX indexes (like a configurations array index) 

BTW, according to the output you are running all the processes on the same node (node name is the same for all outputs)

**Reply by w7658610** [reply]
thank you for answering my question!!
